src/buffer/out/Row.cpp

```diff
@@ -945,20 +1016,33 @@ DbcsAttribute ROW::DbcsAttrAt(til::CoordType column) const noexcept
 
 std::wstring_view ROW::GetText() const noexcept
 {
-    return { _chars.data(), _charSize() };
+    const auto width = size_t{ til::at(_charOffsets, GetReadableColumnCount()) } & CharOffsetsMask;
+    return { _chars.data(), width };
 }
 
 std::wstring_view ROW::GetText(til::CoordType columnBegin, til::CoordType columnEnd) const noexcept
 {
     const til::CoordType columns = _columnCount;
-    const auto colBeg = std::max(0, std::min(columns, columnBegin));
-    const auto colEnd = std::max(colBeg, std::min(columns, columnEnd));
+    const auto colBeg = clamp(columnBegin, 0, columns);
+    const auto colEnd = clamp(columnEnd, colBeg, columns);
     const size_t chBeg = _uncheckedCharOffset(gsl::narrow_cast<size_t>(colBeg));
     const size_t chEnd = _uncheckedCharOffset(gsl::narrow_cast<size_t>(colEnd));
 #pragma warning(suppress : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
     return { _chars.data() + chBeg, chEnd - chBeg };
 }
 
+til::CoordType ROW::GetLeadingColumnAtCharOffset(const ptrdiff_t offset) const noexcept
+{
+    auto mapper = _createCharToColumnMapper(offset);
+    return mapper.GetLeadingColumnAt(offset);

```

Comment:
`_createCharToColumnMapper` only uses the offset to guess the initial column. `GetLeadingColumnAt` will then actually compute the proper offset. This approach avoids `GetLeadingColumnAt` being `O(n)` most of the time.

The reason the `CharToColumnMapper` exists as a separate class is because I want to use it for DirectWrite to map the character positions that DirectWrite returns directly back to columns. Instead, right now, we got a really ugly and inefficient approach. You can see it if you search for `bufferLineColumn`.